Modules/_sqlite/connection.c

```diff
@@ -1442,6 +1457,169 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
     return retval;
 }
 
+#ifdef HAVE_BACKUP_API
+static PyObject *
+pysqlite_connection_backup(pysqlite_Connection *self, PyObject *args, PyObject *kwds)
+{
+    PyObject *target = NULL;
+    const char *filename;
+    int pages = -1;
+    PyObject *progress = Py_None;
+    char *name = "main";
+    int rc;
+    int cberr = 0;
+    double sleep_secs = 0.250;
+    sqlite3 *bckconn;
+    sqlite3_backup *bckhandle;
+    static char *keywords[] = {"target", "pages", "progress", "name", "sleep", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|$iOsd:backup", keywords,
+                                     &target, &pages, &progress, &name, &sleep_secs)) {
+        return NULL;
+    }
+
+    if (!PyUnicode_Check(target) && !PyObject_TypeCheck(target, &pysqlite_ConnectionType)) {
+        PyErr_SetString(PyExc_TypeError,
+                        "target argument must be either a string or a Connection instance");
+        return NULL;
+    }
+
+    if (progress != Py_None && !PyCallable_Check(progress)) {
+        PyErr_SetString(PyExc_TypeError, "progress argument must be a callable");
+        return NULL;
+    }
+
+    if (pages == 0) {
+        pages = -1;
+    }
+
+    if (PyUnicode_Check(target)) {
+        filename = PyUnicode_AsUTF8(target);
+        if (filename == NULL) {
+            return NULL;
+        }
+
+        Py_BEGIN_ALLOW_THREADS
+        rc = sqlite3_open(filename, &bckconn);
+        Py_END_ALLOW_THREADS
+
+        if (rc != SQLITE_OK) {
+            if (bckconn) {
+                _pysqlite_seterror(bckconn, NULL);
+                Py_BEGIN_ALLOW_THREADS
+                SQLITE3_CLOSE(bckconn);
+                Py_END_ALLOW_THREADS
+            } else {
+                (void)PyErr_NoMemory();
+            }
+            return NULL;
+        }
+    }
+    else {
+        if ((pysqlite_Connection *)target == self) {
+            PyErr_SetString(PyExc_ValueError, "target cannot be the same connection instance");
+            return NULL;
+        }
+#if SQLITE_VERSION_NUMBER < 3008007
```

Comment:
3.8.7 was released in 2014, https://www.sqlite.org/releaselog/3_8_7.html

Do we need to support older versions? How about requiring a minimum version of 3.8.7 or even newer to compile CPython?

Reply:
I'm definitely in favor of not supporting old versions of dependencies if it lets us avoid lots of compatibility code.

Reply:
According to distrowatch, Debian Jessie has 3.8.7.1 but CentOS 7-1708 only 3.7.17. Guess we need to support < 3.8 for a while.